tests/pipeline_operations_test.py

```diff
@@ -72,6 +82,57 @@ def test_local_group_by_key(self):
                          ("cheese", ["brie", "swiss"]),
                          ("bread", ["sourdough"])])
 
+    def test_local_filter(self):
+        self.assertEqual(list(self.ops.filter([], lambda x: True)),
+                         [])
+        self.assertEqual(list(self.ops.filter([], lambda x: False)),
+                         [])
+
+        example_list = [1, 2, 2, 3, 3, 4, 2]
+
+        self.assertEqual(list(self.ops.filter(example_list, lambda x: x % 2)),
+                         [1, 3, 3])
+        self.assertEqual(list(self.ops.filter(example_list, lambda x: x < 3)),
+                         [1, 2, 2, 2])
+
+    def test_local_values(self):
+        self.assertEqual(list(self.ops.values([])),
+                         [])
+
+        example_list = [(1, 2), (2, 3), (3, 4), (4, 8)]
+
+        self.assertEqual(list(self.ops.values(example_list)),
+                         [2, 3, 4, 8])
+
+    def test_local_count_per_element(self):
+        example_list = [1, 2, 3, 4, 5, 6, 1, 4, 0, 1]
+        result = self.ops.count_per_element(example_list)
+
+        self.assertEqual(dict(result),
+                         {1: 3, 2: 1, 3: 1, 4: 2, 5: 1, 6: 1, 0: 1})
+
+    def test_laziness(self):
+        def exceptions_generator_function():
+            yield 1 / 0
+
+        def assert_laziness(operator, *args):
+            try:
+                operator(exceptions_generator_function(), *args)
+            except ZeroDivisionError:
+               self.fail(f"local {operator.__name__} is not lazy")
```

Comment:
This function exists for tests to fail properly with `AssertionError` (`assertNotRaises` does not exists in `unittest`). 

e.g. when filter implementation is broken:
```python
        return list(filter(fn, col))
```
the fail looks like this:
```
tests/pipeline_operations_test.py:122: in assert_laziness
    self.fail(f"local {operator.__name__} is not lazy")
E   AssertionError: local filter is not lazy
```

Reply:
Thanks for explanation. I like that approach!